### PR TITLE
Allow creating modules in a module pass

### PR DESCRIPTION
**Pull request by pfultz2**
This adds a `module_pass_manager` class that can be used for module passes. It provides a `create_module` method so module passes can create modules.

This is needed for pointwise ops, so each pointwise op can be fused in a submodule.

### CODE REVIEW

File: tools/include/pass.hpp [hunk 1]
```diff
@@ -24,17 +26,42 @@ struct pass
     /// A unique name used to identify the pass
     std::string name() const;
     /// Run the pass on the module
+    void apply(module_pass_manager& mpm) const;
     void apply(module& m) const;
```
Comment (umangyadav) [inline tools/include/pass.hpp:30]: Same here ? 

Also curious why do we have the same logic both in `src/.../pass.hpp` and `tools/.../pass.hpp` ?

File: src/include/migraphx/pass.hpp [hunk 2]
```diff
@@ -24,20 +26,45 @@ struct pass
     /// A unique name used to identify the pass
     std::string name() const;
     /// Run the pass on the module
+    void apply(module_pass_manager& mpm) const;
     void apply(module& m) const;
```
Comment (umangyadav) [inline src/include/migraphx/pass.hpp:30]: Do we need this method that applies on Module directly ?

File: tools/include/pass.hpp (hunk 1 above)
Comment (pfultz2) [reply tools/include/pass.hpp:30]: > Also curious why do we have the same logic both in src/.../pass.hpp and tools/.../pass.hpp ?

The header under tools/include are ran through te.py to generate the type erasure code. This generated file is then placed under src/include. See [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/wiki/Developer's-Guide#type-erasure).

File: src/include/migraphx/pass.hpp (hunk 2 above)
Comment (pfultz2) [reply src/include/migraphx/pass.hpp:30]: Yes as all of our passes are written using this method.

Comment (umangyadav) [reply src/include/migraphx/pass.hpp:30]: Then, Should all of them now use module_pass_manager instead 
?

Comment (pfultz2) [reply src/include/migraphx/pass.hpp:30]: I dont see a reason to change it.

**Review by scxiao — APPROVED** [review]

**Review by umangyadav — APPROVED** [review]

**Pull request merged**